crates/matrix-sdk/src/encryption/recovery/mod.rs

```diff
@@ -0,0 +1,526 @@
+    /// ```no_run
+    /// # use matrix_sdk::{Client, encryption::backups::BackupState};
+    /// # use url::Url;
+    /// # async {
+    /// # let homeserver = Url::parse("http://example.com")?;
+    /// # let client = Client::new(homeserver).await?;
+    /// use futures_util::StreamExt;
+    ///
+    /// let recovery = client.encryption().recovery();
+    ///
+    /// recovery.enable_backup().await?;
+    ///
+    /// assert_eq!(client.encryption().backups().state(), BackupState::Enabled);
+    ///
+    /// # anyhow::Ok(()) };
+    /// ```
+    pub async fn enable_backup(&self) -> Result<()> {
+        if !self.client.encryption().backups().exists_on_server().await? {
+            self.mark_backup_as_enabled().await?;
+
+            self.client.encryption().backups().create().await?;
+            self.client.encryption().backups().maybe_trigger_backup();
+
+            Ok(())
+        } else {
+            Err(RecoveryError::BackupExistsOnServer)
+        }
+    }
+
+    /// Disable recovery completely.
+    ///
+    /// This method will do the following steps:
+    ///
+    /// 1. Disable the uploading of room keys to a currently active backup.
+    /// 2. Delete the currently active backup.
+    /// 3. Set the `m.secret_storage.default_key` global account data event to
+    ///    an empty JSON content.
+    /// 4. Set a global account data event so clients won't attempt to
+    ///    automatically re-enable a backup.
+    ///
+    /// # Examples
+    ///
+    /// ```no_run
+    /// # use matrix_sdk::{Client, encryption::recovery::RecoveryState};
+    /// # use url::Url;
+    /// # async {
+    /// # let homeserver = Url::parse("http://example.com")?;
+    /// # let client = Client::new(homeserver).await?;
+    /// use futures_util::StreamExt;
+    ///
+    /// let recovery = client.encryption().recovery();
+    ///
+    /// recovery.disable().await?;
+    ///
+    /// assert_eq!(recovery.state(), RecoveryState::Disabled);
+    ///
+    /// # anyhow::Ok(()) };
+    /// ```
+    pub async fn disable(&self) -> Result<()> {
+        self.client.encryption().backups().disable().await?;
+        // Why oh why, can't we delete account data events?
+        self.client.account().set_account_data(SecretStorageDisabledContent {}).await?;
+        self.client.account().set_account_data(BackupDisabledContent { disabled: true }).await?;
+        self.update_recovery_state().await?;
+        // TODO: Do we want to "delete" the known secrets as well?
+
+        Ok(())
+    }
+
+    /// Reset the recovery key.
+    ///
+    /// This will rotate the secret storage key and re-upload all the secrets to
+    /// the [`SecretStore`].
+    ///
+    /// # Examples
+    ///
+    /// ```no_run
+    /// # use matrix_sdk::{Client, encryption::recovery::RecoveryState};
+    /// # use url::Url;
+    /// # async {
+    /// # let homeserver = Url::parse("http://example.com")?;
+    /// # let client = Client::new(homeserver).await?;
+    /// use futures_util::StreamExt;
```

Comment:
… and here.

Reply:
Same 77444856aab8b321070293c3d661a4efa2f33082